oasis-reference/src/main/kotlin/com/empowerops/volition/ref_oasis/DataModelService.kt

```diff
@@ -1,92 +1,176 @@
 
-class DataModelService(
-        val viewData: ViewData,
-        private val updateCallback: () -> Unit
-) : ModelService {
-    override var simByName: Map<String, Simulation> = emptyMap()
+interface Event
+data class StatusUpdateEvent(val status: String) : Event
+data class NewMessageEvent(val message: Message) : Event
+data class NewResultEvent(val result: Result) : Event
+data class PluginRegisteredEvent(val name: String) : Event
+data class PluginUnRegisteredEvent(val name: String) : Event
+class PluginUpdatedEvent : Event
+data class PluginRenamedEvent(val oldName: String, val newName: String) : Event
+data class ProxyAddedEvent(val name: String) : Event
+data class ProxyRemovedEvent(val name: String) : Event
+data class ProxyRenamedEvent(val oldName: String, val newName: String) : Event
+class ProxyUpdatedEvent : Event
+data class IssuesUpdatedEvent(val issues: List<String>) : Event
 
-    override fun updateStatusMessage(message: String) {
-        GlobalScope.launch(Dispatchers.JavaFx) {
-            viewData.currentEvaluationStatus.value = message
-        }
+class DataModelService(private val eventBus: EventBus) {
+    var simByName: Map<String, Simulation> = emptyMap()
+    var proxyByName: Map<String, Proxy> = emptyMap()
+
+    fun updateStatusMessage(message: String) {
+        eventBus.post(StatusUpdateEvent(message))
     }
 
-    override fun addMessage(message: Message) {
-        GlobalScope.launch(Dispatchers.JavaFx) {
-            viewData.allMessages.add(message)
-        }
+    fun addMessage(message: Message) {
+        eventBus.post(NewMessageEvent(message))
     }
 
-    override fun addResult(result: Result) {
-        GlobalScope.launch(Dispatchers.JavaFx) {
-            viewData.resultList.add(result)
-        }
+    fun addResult(result: Result) {
+        eventBus.post(NewResultEvent(result))
     }
 
-    override fun setDuration(nodeName: String?, timeOut: Duration?) {
-        if (nodeName == null || !simByName.containsKey(nodeName)) {
+    private fun updateIssueList() {
+        eventBus.post(IssuesUpdatedEvent(findIssue()))
+    }
+
+    fun setDuration(nodeName: String?, timeOut: Duration?) {
+        if (nodeName == null || nodeName !in proxyByName) {
             return
         }
-        val node: Simulation = simByName.getValue(nodeName)
+        val node = proxyByName.getValue(nodeName)
         val newNode = node.copy(timeOut = timeOut)
-        simByName += nodeName to newNode
+        proxyByName += nodeName to newNode
     }
 
-    override fun renameSim(target: Simulation, newName: String, oldName: String) {
-        require(simByName.containsKey(oldName) && !simByName.containsKey(newName))
-        simByName += newName to target
+    fun renameSim(newName: String, oldName: String): Boolean {
+        if (newName in simByName || oldName !in simByName) {
+            return false
+        }
+        simByName += newName to simByName.getValue(oldName).copy(name = newName)
```

Comment:
hmm. Any reason we dont just use a `list` instead of `simByName` given that `Simulation` has its `name` as a property?

Reply:
It is just easy for me to do a lookup without using filter or map and have access to fuctions for map to easily access key and values. It was just convenient to write it in this way for varies check that I do on adding/removing/renaming. 
I don't think they are different between `nodeName !in myMap` vs  `nodeName !in myList.filter{ it.name }` but I do prefer map code than the list one. 
```
 private fun renameProxy(newName: String, oldName: String): Boolean {
        if (newName in proxyByName.map { it.name } || oldName !in proxyByName.map { it.name }) {
            return false
        }
        val single = proxyByName.single { it.name == oldName }
        proxyByName += single.copy(name = newName)
        proxyByName -= single
        eventBus.post(ProxyRenamedEvent(oldName, newName))
        return true
}
```
vs
```
private fun renameProxy(newName: String, oldName: String): Boolean {
        if (newName in proxyByName || oldName !in proxyByName) {
            return false
        }
        proxyByName += newName to proxyByName.getValue(oldName).copy(name = newName)
        proxyByName -= oldName
        eventBus.post(ProxyRenamedEvent(oldName, newName))
        return true
}
```
Now, since we do have names in the node, I agree we should do either list with name and map with name.

Reply:
And maybe I am play around more about null safe call here.

Reply:
We chatted about this and this is good exploration and learning opportunity for me as I moved to a list and playing around with extension functions.
 
I did my changes, not really liking it in general. To me, it is just more readable with a map in most of the case but I still don't think they are big difference. 
Checking the map/list already have the name:
For map: `name in map.key `
For list: `list.singleOrNull{name == it.name}` or `name in list.map{it.name}`
Rettving a node is just need more code: non nullable `map[name]` vs `list.single{it.name == name}` and nullable version which use `getValue` vs `singleOrDefault`. Also, in some of the lambda block, I have to use specify the lambda parameter since we lost it.
```
{ name ->
   dataModel.simulations.single{it.name == name}.stuff
} 
```
vs
```
{
  map.getValue(it).stuff
}
```
For update, we do have to make sure to remove the old node, where map for the same key, to update you just need to add them again with the same key and you need to worry about it when rename happened. 

ALl that being said, I don't have problem with this, since it just different logic and map have more dedicated feature for these. And this is easily solvable via extension functions,for example use `List<Simulation>.getValue`. I am just a bit surprised how big of a difference there actually are. I am going to keep this after I play around with extension functions.